### DISCUSSION TITLE
Adding Role Definitions

### DISCUSSION THREAD

**Discussion by kentlhastings** [pnp/pnpcore#369] [Q&A] [answered]
Hello all,

I am trying to define specific RoleDefinitions for multiple folders and files.

Given the folder or file's serverRelativeUrl and a set of PrincipalIds and string[]s containing the permission names, I want to do something like this:

var source = context.Web.GetFolderByServerRelativeUrl(path, x => x.ItemCount, x => x.Folders, x => x.Files, x => x.Parent, x => x.ListItemAllFields);

foreach (var permission in permissions)
{
source.ListItemAllfields.AddRoleDefinitions(permission.Key, permission.Value);
}

This code is failing on the AddRoleDefinitions() call however. Does anyone have any suggestions about making it work?

Thank you!

**Comment by jansenbe** [comment, marked as answer]
@kentlhastings : I think the AddRoleDefinitions call failed when using ListItemAllFields (just like BreakRoleInheritance)...I've just fixed this, so tomorrow things should work. If not, please create an issue in the issue list.

**Reply by kentlhastings** [reply]
This particular call is failing in a different place. There is a line where Parent.Parent is cast as a List. So the value derived from this is null, but the query string wants the Id property, so we get a null reference exception.

**Reply by kentlhastings** [reply]
@jansenbe, it's failing right here:

public async Task<bool> AddRoleDefinitionsAsync(int principalId, params string[] names)
        {
            var parentList = Parent.Parent as List;
            var result = false;
            foreach (var name in names)
            {
                var roleDefinition = await PnPContext.Web.RoleDefinitions.FirstOrDefaultAsync(d => d.Name == name).ConfigureAwait(false);
                if (roleDefinition != null)
                {
                    var apiCall = new ApiCall($"_api/web/lists(guid'{parentList.Id}')/Items({Id})/roleassignments/addroleassignment(principalid={principalId},roledefid={roleDefinition.Id})", ApiType.SPORest);
                    await RawRequestAsync(apiCall, HttpMethod.Post).ConfigureAwait(false);
                    return true;
                }
                else
                {
                    throw new ArgumentException($"Role definition '{name}' not found.");
                }
            }
            return result;
        }

**Reply by jansenbe** [reply]
This is the new code:
![image](https://user-images.githubusercontent.com/7451219/112531583-0900c680-8da8-11eb-913f-5c2fb6cb30db.png)

**Reply by kentlhastings** [reply]
Awesome!! Thanks so much!

**Reply by onoflo** [reply]
Sorry, but what are the principalId and the names params?
